glean-core/rlb/src/net/mod.rs

```diff
@@ -0,0 +1,55 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! A module responsible for the Glean upload logic.
+//!
+//! This doesn't perform the actual upload but rather handles
+//! retries, upload limitations and error tracking.
+
+use crate::with_glean;
+pub use glean_core::upload::{PingRequest, UploadResult};
+use glean_core::upload::PingUploadTask;
+
+pub trait PingUploader {
+    // UploadResult comes from glean-core.
+   fn upload(&self, request: &PingRequest) -> UploadResult;
+}
+
+struct BaseUploader<T> {
+   uploader: T,
+}
+
+impl<T> BaseUploader<T> where T: PingUploader {
+    fn new(new_uploader: T) -> Self {
+        Self {
+           uploader: new_uploader
+        }
+    }
+   
+    pub(crate) fn trigger_upload(&self) {
+        loop {
+            let incoming_task = with_glean(|glean| glean.get_upload_task());
+
+            match incoming_task {
+                PingUploadTask::Upload(request) => {
+                    let doc_id = request.document_id.clone();
+                    let result = self.uploader.upload(&request);
+                    // Process the upload response.
+                    with_glean(|glean| {
+                        // TODO: this works but it's not ideal
+                        //glean.process_ping_upload_response(&doc_id, self.uploader.upload(&request))
+                        glean.process_ping_upload_response(&doc_id, result)
```

Comment:
```
error[E0507]: cannot move out of result, a captured variable in an Fn closure
--> glean-core\rlb\src\net\mod.rs:39:84
|
37 | let result = self.uploader.upload(&request);
| ------ captured outer variable
38 | // Process the upload response.
39 | with_glean(|glean| glean.process_ping_upload_response(&doc_id, result));
| ^^^^^^ move occurs because result has type glean_core::upload::UploadResult, which does not implement the Copy trait

error[E0507]: cannot move out of result, a captured variable in an Fn closure
--> glean-core\rlb\src\net\mod.rs:39:84
|
37 | let result = self.uploader.upload(&request);
| ------ captured outer variable
38 | // Process the upload response.
39 | with_glean(|glean| glean.process_ping_upload_response(&doc_id, result));
| ^^^^^^ move occurs because result has type glean_core::upload::UploadResult, which does not implement the Copy trait

error: aborting due to previous error; 2 warnings emitted

For more information about this error, try rustc --explain E0507.
error: aborting due to previous error; 2 warnings emitted

For more information about this error, try rustc --explain E0507.
error: could not compile glean.

To learn more, run the command again with --verbose.
warning: build failed, waiting for other jobs to finish...
error: build failed

```

@badboy @brizental , this puzzles me a bit. I think I understand what's going on: `result` has a different lifetime than the scope within `with_glean` and its ownership cannot be transferred.

If I uncomment the previous line and comment this one, everything works just fine. However, that would not be ideal: I don't want to hold `glean` while doing networking stuff.

Moreover, I see that [FOG seems to do it right](https://searchfox.org/mozilla-central/rev/c409dd9235c133ab41eba635f906aa16e050c197/toolkit/components/glean/api/src/ping_upload.rs#47), but I can't understand why that works. Is it because of the `Send`?

Reply:
Did you try with `with_glean(move |glean| { ... }`?
(Otherwise I need to take a closer look later)

Reply:
Yes, i did, it doesn't help with that error apparently.

Reply:
Ah, `with_glean` takes `F: Fn` in Rlb, whereas it takes `F: FnOnce` in FOG: https://searchfox.org/mozilla-central/source/toolkit/components/glean/api/src/lib.rs#29-31

The latter is what we want here to give the most freedom.

Reply:
Thanks! That was it!